key-share/src/valid.rs

```diff
@@ -153,6 +174,9 @@ pub trait ValidateFromParts<Parts>: Validate {
     fn from_parts(parts: Parts) -> Self;
 }
 
+/// Marker trait stating that `Self` is projection of `T`: if `T` is valid, then `Self::from(T)` is valid as well
+pub trait ValidProjection<T: Validate>: Validate + From<T> {}
+
```

Comment:
It seems we want to have the instances for cggmp keyshare+aux too?

Reply:
What do you mean exactly?

Reply:
Ah probably you mean that we should add
```rust
impl ValidProjection<KeyShare> for IncompleteKeyShare {}
impl ValidProjection<KeyShare> for AuxData {}
```

that does make sense

Reply:
Tho now I see a pattern, probably what we want is ValidationLens? Wdyt?

```rust
/// If `Self` contains `T` and `Self` was validated, then `T` is also valid
pub trait ValidationLens<T> {
    fn lens(&self) -> &T;
}
```

Reply:
and then 

```rust
impl<T: Validate> Valid<T> {
    pub fn lens<L>(&self) -> Valid<&L>
    where
        T: ValidationLens<L>
    { /* ... */ }
}

Reply:
naming is confusing tho

Reply:
As for lens, I think From is more useful. For example, in key-share we have (or will have) functions that accept `CoreKeyShare ~ Valid<DirtyCoreShare>`, but we want to call them somehow with `KeyShare ~ Valid<DirtyKeyShare>`. Thus we need conversion `KeyShare -> CoreKeyShare`.

Lens instead gives us `&KeyShare -> Valid<&DirtyKeyShare>` which is useless. We could recover some of the usefulness with `cloned :: Valid<&T> -> Valid<T>`, but this just rediscovers From::from again

Reply:
Though I agree that the function `project :: forall T S. ValidProjectionFrom<T> S => Valid<T> -> Valid<S>` must exist; I just tried to write some conversions without it and failed

Reply:
I replaced the whole thing with AsRef trick